File: tutorials/vbnet/hello_world/win_async/AsyncTest.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports SocketProAdapter
Imports SocketProAdapter.ClientSide

Namespace win_async
    Partial Public Class AsyncTest
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Function GetTask() As Task(Of String)
            Dim hw As HelloWorld = m_spHw.AsyncHandlers(0)
            Return hw.Async(Of String, String, String)(hwConst.idSayHelloHelloWorld, "Jack", "Smith")
        End Function

        Private Function GetTasksInBatch() As Task(Of String)
            Dim hw As HelloWorld = m_spHw.AsyncHandlers(0)
            Dim ok As Boolean = hw.SendRequest(hwConst.idSleepHelloWorld, 5000, Sub(ar)
                                                                 End Sub)
            Dim task As Task(Of String) = hw.Async(Of String, String, String)(hwConst.idSayHelloHelloWorld, "Jone", "Don")
            Return task
        End Function

        Private m_spHw As New CSocketPool(Of HelloWorld)(True)

        Private Sub async_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            Dim cc As New CConnectionContext("127.0.0.1", 20901, "MyUserId", "MyPassword")
            btnTest.Enabled = m_spHw.StartSocketPool(cc, 1, 1)
        End Sub

        Private Async Sub btnTest_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnTest.Click
            If m_spHw.ConnectedSockets = 0 Then
                txtRes.Text = "No connection"
                Return
            End If
            btnTest.Enabled = False
            Try
                'execute one request asynchronously
                txtRes.Text = Await GetTask()

                'execute multiple requests asynchronously in batch
                txtRes.Text = Await GetTasksInBatch()
                btnTest.Enabled = True
            Catch err As Exception
                txtRes.Text = err.Message
            End Try
        End Sub
    End Class
End Namespace

File: tutorials/csharp/hello_world/win_async/AsyncTest.cs
using System;
using System.Threading.Tasks;
using System.Windows.Forms;
using SocketProAdapter;
using SocketProAdapter.ClientSide;

namespace win_async
{
    public partial class AsyncTest : Form
    {
        public AsyncTest()
        {
            InitializeComponent();
        }

        Task<CScopeUQueue> GetTask()
        {
            HelloWorld hw = m_spHw.AsyncHandlers[0];
            return hw.Async(hwConst.idSayHelloHelloWorld, "Jack", "Smith");
        }

        Task<CScopeUQueue> GetTasksInBatch()
        {
            HelloWorld hw = m_spHw.AsyncHandlers[0];
            bool ok = hw.SendRequest(hwConst.idSleepHelloWorld, 5000, (ar) => { });
            Task<CScopeUQueue> task = hw.Async(hwConst.idSayHelloHelloWorld, "Jone", "Don");
            return task;
        }

        CSocketPool<HelloWorld> m_spHw = new CSocketPool<HelloWorld>(true);

        private void async_Load(object sender, EventArgs e)
        {
            CConnectionContext cc = new CConnectionContext("127.0.0.1", 20901, "MyUserId", "MyPassword");
            btnTest.Enabled = m_spHw.StartSocketPool(cc, 1);
        }

        private void AsyncTest_FormClosed(object sender, System.Windows.Forms.FormClosedEventArgs e)
        {
            //Crash may happen randomly if there is a request in processing! You can use the following logic to avoid the issue.
            if (m_spHw.Sockets[0].CountOfRequestsInQueue > 0)
            {
                //Optionally force to wait until all requests are processed
                //m_spHw.Sockets[0].WaitAll();

                //explicitly shutdown pool and ignore all requests in processing
                m_spHw.ShutdownPool();
            }
        }

        private async void btnTest_Click(object sender, EventArgs e)
        {
            if (m_spHw.ConnectedSockets == 0)
            {
                txtRes.Text = "No connection";
                return;
            }
            btnTest.Enabled = false;
            try
            {
                //execute one request asynchronously
                txtRes.Text = (await GetTask()).Load<string>();

                //execute multiple requests asynchronously in batch
                txtRes.Text = (await GetTasksInBatch()).Load<string>();
                btnTest.Enabled = true;
            }
            catch (Exception err)
            {
                txtRes.Text = err.Message;
            }
        }
    }
}
